File: src/main/python/com/hazza/LeetCodeByPython/SegmentOne/No45_Jump_Game_II.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/5/16 20:31
# https://leetcode.com/problems/jump-game-ii/
from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:
        steps, cur, curMax = 0, 0, 0
        numsLen = len(nums)

        for i, n in enumerate(nums[:-1]):
            curMax = max(curMax, i + n)
            if i == cur:
                steps += 1
                cur = curMax
                if cur >= numsLen:
                    break

        return steps


if __name__ == '__main__':
    solution = Solution()

    print(solution.jump([2, 3, 1, 1, 4]))

    assert solution.jump([2, 3, 1, 1, 4]) == 2

File: src/main/java/com/hazza/leetcode/SegmentOne/No45JumpGameII.java
package com.hazza.leetcode.SegmentOne;

/**
 * https://leetcode.com/problems/jump-game-ii/description/
 */
public class No45JumpGameII {
    //solution 1
    public int jump(int[] nums) {
        if (nums == null || nums.length < 2) return 0;
        int len = nums.length, preMax = 0, curMax= 0, steps = 0;
        for (int i = 0; i < len - 1; i++) {
            curMax = Math.max(curMax, nums[i] + i);
            if (i == preMax) {
                ++steps;
                preMax = curMax;
            }
        }

        return steps;
    }

    //solution 2
    public int jump2(int[] nums) {
        if (nums == null || nums.length < 2) return 0;
        int len = nums.length;
        int curMax = 0, nextMax = 0, steps = 0, i = 0;
        while (curMax - i + 1 > 0) {
            ++steps;
            while (i <= curMax) {
                nextMax = Math.max(nextMax, nums[i] + i);
                if (nextMax >= len - 1) return steps;
                ++i;
            }
            curMax = nextMax;
        }
        return 0;
    }
}
